dashboard/test/lib/services/curriculum_pdfs/lesson_plans_test.rb

```diff
@@ -20,19 +20,19 @@ class Services::CurriculumPdfs::LessonPlansTest < ActiveSupport::TestCase
 
   test 'urls are escaped' do
     script = create(:script, name: "test-escapes-script", seeded_from: Time.at(0))
-    lesson = create(:lesson, script: script, key: "Some!key_with?special/characters")
```

Comment:
It looks like, theoretically, two lessons could end up with the same canonicalized filename (ie `Some!key_with!special!character` and `Some?key_with?special?character` would both have the filename `Some-name_with-special-characters.pdf`). This seems like a reasonable limitation to me as it would only happen with special characters in the keys, but wanted to double check we don't have lessons within the same script that would fall into this now?

Reply:
correct! yes, it is not ideal, but seems like a reasonable tradeoff to me. Confirmed, there are no collisions:
```
Services::CurriculumPdfs.get_pdf_enabled_scripts.each do |script|
  h = {}
  script.lessons.each do |l|
    raise if h[l.name.parameterize]
    h[l.name.parameterize] = 1
  end
end
```

one tangential takeaway is that I could be using lesson.name, which is the same as localized_name but cheaper to calculate.

If we wanted to truly solve this, one idea would be to (1) go back to using lesson key instead of lesson name, and (2) start canonicalizing lesson key everywhere (in the database).

Reply:
I'd love to do the second piece but that's way out of scope for this :) I'm happy knowing that there are no collisions as the chance of one is low